commands/bootstrap

```diff
@@ -297,6 +332,331 @@ create_self_signed() {
+        info "Using existing private key"
+    fi
+
+    # Create Certificate Signing Request (CSR)
+    DEVICE_CSR=$(
+        "${EXEC_CMD[@]}" "$TARGET" $SUDO openssl req \
+            -key "$LOCAL_CERT_KEY" \
+            -new \
+            -subj "/O=thin-edge/OU=${DEVICE_ID}/CN=${SERVICE}"
+    )
+
+    # Sign the CSR
+    if [ "$SERVICE" = "child-agent" ]; then
+    CERT_EXT=$(cat << EOF
+basicConstraints=CA:FALSE
+extendedKeyUsage = clientAuth
+EOF
+    )
+    else
+    CERT_EXT=$(cat << EOF
+authorityKeyIdentifier=keyid
+basicConstraints=CA:FALSE
+keyUsage = digitalSignature, keyAgreement
+extendedKeyUsage = serverAuth, clientAuth
+subjectAltName = DNS:${TARGET_DNS},IP:${TARGET_IP},DNS:localhost,IP:127.0.0.1
+EOF
+    )
+    fi
+
+    DEVICE_CERT_CHAIN="/tmp/${LOCAL_CERT_NAME}.crt"
+    openssl x509 -req \
+        -in <(echo "$DEVICE_CSR") \
+        -out "$DEVICE_CERT_CHAIN" \
+        -CA "$SIGNING_CERT_FILE" \
+        -CAkey "$SIGNING_CERT_KEY" \
+        -extfile <(echo "$CERT_EXT") \
+        -CAcreateserial \
+        -days "$LEAF_EXPIRE_DAYS"
+
+    # Build certificate chain (from leaf cert to the signing cert)
+    cat "$SIGNING_CERT_FILE" >> "$DEVICE_CERT_CHAIN"
+
+    # Copy cert to device
+    "${SCP_CMD}" "$DEVICE_CERT_CHAIN" "$TARGET" "/tmp/${LOCAL_CERT_NAME}.crt"
+    "${EXEC_CMD[@]}" "$TARGET" "$SUDO" mv "/tmp/${LOCAL_CERT_NAME}.crt" "$LOCAL_CERT_FILE"
+    "${EXEC_CMD[@]}" "$TARGET" "$SUDO" chown tedge:root "$LOCAL_CERT_FILE"
+    "${EXEC_CMD[@]}" "$TARGET" "$SUDO" chmod 644 "$LOCAL_CERT_FILE"
+
+    case "$SERVICE" in
+        mqtt-broker)
+            configure_mtls_mqtt_broker "$TARGET_IP" "$TARGET_DNS" "$LOCAL_SIGNING_CERT_FILE" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
+            ;;
+        main-agent)
+            configure_mtls_mqtt_client "$TARGET_IP" "$TARGET_DNS" "$LOCAL_SIGNING_CERT_FILE" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
+            configure_mtls_main_agent "$TARGET_IP" "$TARGET_DNS" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
+            ;;
+        c8y-mapper)
+            configure_mtls_c8y_mapper "$TARGET_IP" "$TARGET_DNS" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
+            ;;
+        child-agent)
+	    # GET IP and DNS of the main device
+	    MAIN_DEVICE_DNS=$("${EXEC_CMD[@]}" "$MAIN_DEVICE" hostname 2>/dev/null || true)
+            if [ -z "$MAIN_DEVICE_DNS" ]; then
+                MAIN_DEVICE_DNS=$(echo "$MAIN_DEVICE" | sed -E 's/.*@(\S+)/\1/')
+            fi
+            MAIN_DEVICE_IP=$("${EXEC_CMD[@]}" "$MAIN_DEVICE" ip route get 8.8.8.8 2>/dev/null | sed -E 's/.*src (\S+) .*/\1/;t;d' 2>/dev/null || true)
+            if [ -z "$MAIN_DEVICE_IP" ]; then
+                MAIN_DEVICE_IP=$(echo "$MAIN_DEVICE" | sed -E 's/.*@(\S+)/\1/')
+            fi
+            configure_mtls_mqtt_client "$MAIN_DEVICE_IP" "$MAIN_DEVICE_DNS" "$LOCAL_SIGNING_CERT_FILE" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
+            configure_mtls_child_agent "$MAIN_DEVICE_IP" "$MAIN_DEVICE_DNS" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
```

Comment:
Do we need to first check if the child device is actually running the `tedge-agent` service? Otherwise we could extend it in a follow up PR to include the installation of the tedge-agent on the child device as well (though this will probably require a change to the thin-edge.io/install.sh script first)

Reply:
We won't touch any services on the child device for now. I'll create a ticket in the thin-edge.io project to change the install.sh script to also support installing a reduced set of components on a child device (e.g. only the tedge-agent).